subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java

```diff
@@ -901,28 +903,54 @@ public TaskDependency getMustRunAfter() {
 
     @Override
     public void setFinalizedBy(final Iterable<?> finalizedByTasks) {
-        taskMutator.mutate("Task.setFinalizedBy(Iterable)", new Runnable() {
-            @Override
-            public void run() {
-                finalizedBy.setValues(finalizedByTasks);
-            }
+        taskMutator.mutate("Task.setFinalizedBy(Iterable)", () -> {
+            finalizedBy.clear();
+            finalizedBy.add(newFinalizer(ImmutableSet.copyOf(finalizedByTasks), Runnables.doNothing()));
         });
     }
 
     @Override
     public Task finalizedBy(final Object... paths) {
-        taskMutator.mutate("Task.finalizedBy(Object...)", new Runnable() {
-            @Override
-            public void run() {
-                finalizedBy.add(paths);
-            }
+        taskMutator.mutate("Task.finalizedBy(Object...)", () -> {
+            finalizedBy.add(newFinalizer(ImmutableSet.of(paths), Runnables.doNothing()));
         });
         return this;
     }
 
+    @Override
+    public <T extends Buildable> void finalizedBy(T finalizer, Action<T> action) {
+        taskMutator.mutate("Task.finalizedBy(Buildable, Action)", () -> {
+            finalizedBy.add(newFinalizer(ImmutableSet.of(finalizer), () -> action.execute(finalizer)));
+        });
+    }
+
+    private Finalizer newFinalizer(ImmutableSet<Object> dependencies, Runnable action) {
+        TaskDependencyFactory taskDependencyFactory = project.getTaskDependencyFactory();
+        DefaultTaskDependency defaultTaskDependency = taskDependencyFactory.configurableDependency(dependencies);
+
+        return new Finalizer() {
```

Comment:
💭 We shouldn't rely on an anonymous type for this, `DefaultFinalizer` (or something similar) should also be an `abstract` top-level type, instantiated through an instantiator.  Not important for a rough spike, just leaving comments for later refinement.